docs/proposals/2022-07-13-appset-progressive-rollout-strategy.md

```diff
@@ -0,0 +1,175 @@
+
+#### "Pausing" Application Changes During Rollout
+To implement the “paused” functionality of Applications that are not yet ready to be updated, we have a few options.
+* Disable auto-sync.
+** Potentially conflicts with user provided auto-sync settings.
+** Provides the benefit of being able to see the full diff of the ApplicationSet change.
+* “Pause” the Application.
+** Not Yet Implemented: https://github.com/argoproj/argo-cd/issues/4808
+* Prevent any updates at all to the live Applications via the rolling update strategy defined.
+** This is likely the initial implementation method we'll target.
+
+#### Draft Pull Request
+This PR is now functional and ready for comment. We are actively working on unit tests and documentation.
+https://github.com/wmgroot/argo-cd/pull/1
+
+### Security Considerations
+We do not believe this proposal results in any new security considerations for the ApplicationSet controller.
+
+### Risks and Mitigations
+
+If this proposal is implemented, I believe the next logical step would be to solve the case where users would like to control rollout order for Application resources with a consistent specification, but changes being pushed to the upstream `source` of the Application. A common use case is an update to an unversioned "wrapper" helm chart that depends on a versioned upstream chart. The wrapper chart is often used to apply simple supplementary resources in a gitops pattern, such as company specific RBAC configuration, or ExternalSecrets configuration. These supplementary resources do not typically warrant publishing a versioned wrapper chart, making it difficult to implement changes to the chart's templates or value files and roll them out in an ordered way with the ApplicationSet changes discussed here.
+
+Implementing progressive rollout stragies to handled changes upstream of the generated Application source could be difficult, since the applicationset controller would need to intercept the sync operation of the Application to prevent the changes from syncing automatically.
+
+Added maintenance burden on the ArgoCD team is always a risk with the addition of new features.
+
+### Upgrade / Downgrade Strategy
+
+We are introducing new fields to the ApplicationSet CRD, however no existing fields are being changed. We believe this means that a new ApplicationSet version is unnecessary, and that upgrading to the new spec with extra fields will be a clean operation.
+
+Downgrading would risk users receiving K8s API errors if they continue to try to apply the `strategy` field to a downgraded version of the ApplicationSet resource.
+Downgrading the controller while keeping the upgraded version of the CRD should cleanly downgrade/revert the behavior of the controller to the previous version without requiring users to adjust their existing ApplicationSet specs.
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
```

Comment:
There is some prior art here: https://github.com/Skyscanner/applicationset-progressive-sync